### PR TITLE
Cross validation

### PR DESCRIPTION
**Pull request by lbybee**
This pull request adds a cross-validation method `fit_path`, which fits an entire path of IPCA models for each regularizing constant.  It then returns a matrix containing the MSE for each CV partition.  `fit_path` has the option to be run in parallel (across regularizing constants).

Note that I don't want to merge this quite yet as there are several more changes I'd like to make.   However, there are some points that seemed worth discussing to start.

The main point is that `IPCARegressor` now inherits from sklearn's `BaseEstimator`.   The reason for this is that if we can align `IPCARegressor` with the estimator architecture from sklearn we can leverage a lot of their code base and likely make the method more accessible to outside users.

One concrete example would be using their cross-validation framework instead of having our own method.

See here for some docs on what is required to align an estimator with sklearn:

https://scikit-learn.org/dev/developers/contributing.html#rolling-your-own-estimator

The main thing that this would change is the handling of `Panel`.  We may want to do something about that anyways though (see my note on `predict` below).  Currently, I'm thinking it makes sense to split the returns out from `Panel` into their own input, fit would then be something like:

`fit(chars=None, rets=None, PSF=None)`

where `chars` is what `Panel` originally was (without the return column).

Couple of other minor changes this PR makes:

1. I added the singularity check from `_BS_Walpha_sub` to `_BS_Wbeta_sub` as well.
2. I moved the regularized fit params (`alpha`, `l1_ratio`) into the initial class init to align better with other params/sklearn's fit approach.

There are also several other changes I'm considering before I merge:

1. Could likely do some sort of "hot-start" where the final values from the previous regularizing constant are used as starting values in the current iteration.  

    To handle this I'll probably add some optional key-words to `fit`, `Gamma_Old` and `Factor_Old` which will default to the current initialization procedure if `None`.

    This would also require that the parallelization is done within the CV loop but I think that should be fine here.

2. Currently, the `Panel` which `predict` takes as input differs from the `Panel` used elsewhere.  This leads to the somewhat hacky behavior where we do:

    `IPCA.predict(np.delete(Panel, 2, axis=1))`

If we split the returns out this wouldn't be an issue any longer.

3. If I redo how `Panel` is handled, I'll likely remove the `refit` check since it shouldn't be necessary anymore.

### CODE REVIEW

**Comment by matbuechner** [comment]
> The main point is that `IPCARegressor` now inherits from sklearn's `BaseEstimator`. The reason for this is that if we can align I`PCARegressor` with the estimator architecture from sklearn we can leverage a lot of their code base and likely make the method more accessible to outside users.

Making the package more accessible to Scikit-learn users via inheritance from their base estimator is a great idea! In particular, we could think about using some of their unit tests in our testing routines as well. If you have suggestions regarding more natural tests, let me know.

> The main thing that this would change is the handling of Panel. We may want to do something about that anyways though (see my note on predict below). Currently, I'm thinking it makes sense to split the returns out from Panel into their own input, fit would then be something like:

> `fit(chars=None, rets=None, PSF=None)`

> where `chars` is what Panel originally was (without the return column).

Regarding the above, I agree as well. My suggestion would be to use parameter names that are more general, though, to emphasise that IPCA can be employed to data other than returns and asset characteristics. Maybe something like `fit(cond=None, dep=None, psf=None)` where `cond` stands for conditioning information and `dep` for dependent variable. 

> Currently, the Panel which `predict` takes as input differs from the `Panel` used elsewhere. This leads to the somewhat hacky behavior where we do:

> `IPCA.predict(np.delete(Panel, 2, axis=1))`

I agree, the behaviour of predict is not optimal. Your suggests solution that results from keeping characteristics and dependent variable separate sounds plausible to me.  We should also investigate  vectorizing `predict` looping over the dates as is implemented currently is most certainly not optimal.

**Comment by matbuechner** [comment]
> Couple of other minor changes this PR makes:

> I added the singularity check from `_BS_Walpha_sub` to` _BS_Wbeta_sub` as well.

By the way, I add this to `master` already in an earlier commit.

**Comment by lbybee** [comment]
> Regarding the above, I agree as well. My suggestion would be to use parameter names that are more general, though, to emphasise that IPCA can be employed to data other than returns and asset characteristics. Maybe something like fit(cond=None, dep=None, psf=None) where cond stands for conditioning information and dep for dependent variable.

sklearn uses `X` and `y` for this, we may need to follow their convention anyways to align with the interface (not 100% sure on this).

Regardless, sounds like we're on the same page so I'll add the changes to align with sklearn more completely.

**Comment by matbuechner** [comment]
Just a thought on the cross-validation procedure. Could we just use the cross-validated variant of ElasticNet in Sklearn? This would directly return the best estimator along the path of hyper-parameters and save some coding.

[https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNetCV.html](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNetCV.html)

**Comment by lbybee** [comment]
The problem with that approach is that we fit a regularized model for each ALS Gamma step.  We don't necessarily want to select the best model at each ALS step (since this has considerable path dependency with the current factor estimates), we'd rather select the best model _overall_.

That said, using Sklearn's built in cross-validation is, I think, a better approach than what we're currently doing (which does leverage their train/test sample generator but not the full infrastructure).  This is a big part of why I started thinking about adapting IPCA to an sklearn estimator.  Once we've done that we'll have an estimator (comparable to `ElasticNet`) which could have its own variant -- `IPCARegressorCV` or something similar.

**Comment by lbybee** [comment]
Ok, this should be ready to merge.  I have some more thoughts on things to do, but this PR is already pretty far beyond the original scope so I'm going to put those in a separate issue.  Key points to note:

1. Panel input has been split into `X` and `y`.
2. I've added/updated most of the methods to align with sklearn, though we're still not quite there.  The major road-block right now is the panel indices in `X` (time/assets).  sklearn wants `X` to just be data so in order to fully leverage their test suites and what not we'll need to handle this (will say more in the issue).
3. I've added a `score` method, which essentially replaces the `_R2_comps` method.  This aligns better with sklearn (is necessary for some of their framework) and is somewhat more flexible, because you can now get the separate R2 components independently.
4. For most methods there is a `data_type` param which sets whether we use the panel data or the portfolio data (this applies to `fit`, `predict` and `score`).  Currently `fit` defaults to portfolio data because that is faster, but we'll fall back to the panel when doing the regularized estimation.  The `predict` and `score` methods default to the panel data since this seems more intuitive and aligns better with other code bases.
5. For the actual cross-validation, I've kept the `fit_path` method for the time being.  Long-term, I'd like to add an `IPCARegressorCV` class but we'll need to handle the panel index issue before that fully makes sense.

@matbuechner since there are a decent number of changes in here, let me know if you're cool with everything and then I'll merge.

**Comment by lbybee** [comment]
One other change I'll note, `fit` now returns itself (again, to align with sklearn).

File: ipca/ipca.py [hunk 1]
```diff
@@ -71,27 +89,33 @@ def fit(self, Panel=None, PSF=None, refit=False, alpha=0., l1_ratio=1.,
...
-        alpha : scalar
-            Regularizing constant for Gamma estimation.  If this is set to
-            zero then the estimation defaults to non-regularized.
+        data_type : str
+            label for data-type used for ALS estimation, can be one of:
 
-        l1_ratio : scalar
-            Ratio of l1 and l2 penalties for elastic net Gamma fit.
+            1. panel
+            2. portfolio
```
Comment (matbuechner) [inline ipca/ipca.py]: To make the docstring more self-explanatory, can we specify more clearly what we mean by portfolio?
